src/operator/numpy/linalg/np_norm-inl.h

```diff
@@ -0,0 +1,830 @@
+    temp_flat = temp_flat * broadcast_to(trans_in_grad, temp.shape_);
+    if (req[0] == kAddTo) {
+      TBlob ograd(reinterpret_cast<DType*>(tempM.dptr_), outputs[0].shape_,
+                  workspace.dev_mask(), workspace.dev_id());
+      TransposeImpl<xpu>(ctx.run_ctx, temp.reshape(svd_in_shape), ograd, reduce_axes);
+      Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
+      out += ograd.FlatTo1D<xpu, DType>(s);
+    } else {
+      TransposeImpl<xpu>(ctx.run_ctx, temp.reshape(svd_in_shape), outputs[0], reduce_axes);
+    }
+  });
+  if (!param.keepdims) {
+    const_cast<std::vector<TBlob>&>(inputs)[0] = inputs[0].reshape(old_shape_);
+    const_cast<std::vector<TBlob>&>(inputs)[5] = inputs[5].reshape(old_shape_);
+  }
+}
+
+template<typename xpu>
+void NumpyNormComputeForward(const nnvm::NodeAttrs& attrs,
+                             const OpContext& ctx,
+                             const std::vector<TBlob>& inputs,
+                             const std::vector<OpReqType>& req,
+                             const std::vector<TBlob>& outputs) {
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  if (inputs[0].shape_.Size() == 0U) {
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      mxnet::op::Fill<false, DType, xpu>(s, outputs[0], req[0], DType(0.0));
+    });
+    return;
+  }
+  const NumpyNormParam& param = nnvm::get<NumpyNormParam>(attrs.parsed);
+
+  if (param.flag == -2) {  // flattened L2 norm
+    std::vector<TBlob> flat_inputs({
+      inputs[0].reshape(TShape(1, inputs[0].shape_.Size()))
+    });
+    std::vector<TBlob> flat_outputs({
+      outputs[0].reshape(TShape(1, 1))
+    });
+    ReduceAxesComputeImpl<xpu, mshadow_op::nrm2, false, false, mshadow_op::identity>(
+      ctx, flat_inputs, req, flat_outputs, TShape(1, 1));
+    return;
+  }
+
+  if (param.axis.value().ndim() == 2) {
+    NumpyMatrixNormCompute<xpu>(attrs, ctx, inputs, req, outputs);
+  } else {
+    NumpyLpNormCompute<xpu>(attrs, ctx, inputs, req, outputs);
+  }
+}
+
+template<typename xpu>
+void NumpyNormComputeBackward(const nnvm::NodeAttrs& attrs,
+                              const OpContext& ctx,
+                              const std::vector<TBlob>& inputs,
+                              const std::vector<OpReqType>& req,
+                              const std::vector<TBlob>& outputs) {
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  if (inputs[0].shape_.Size() == 0U) {
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      mxnet::op::Fill<false, DType, xpu>(s, outputs[0], req[0], DType(0.0));
+    });
+    return;
+  }
+  const NumpyNormParam& param = nnvm::get<NumpyNormParam>(attrs.parsed);
+
+  if (param.flag == -2) {  // flattened L2 norm
+    std::vector<TBlob> flat_inputs({
+      inputs[0].reshape(TShape(1, 1)),
+      inputs[4].reshape(TShape(1, outputs[0].shape_.Size())),
+      inputs[5].reshape(TShape(1, 1))
+    });
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
```

Comment:
I think we should not allow computation of gradients of integer inputs, so one thing is that we can use `MXNET_INT_TYPE_SWITCH` here instead, another thing is that we can write it like this to provide a better error message:
```c++
if (common::is_float(outputs[0].type_flag_)) {
  // your gradient implementation
} else {
  LOG(FATAL) << "Computing gradient for integer inputs is not well-undefined behavior";
}
```